pkg/operator/discover/discover.go

```diff
@@ -0,0 +1,217 @@
+		return fmt.Errorf("failed to init RBAC for rook-discover. %+v", err)
+	}
+
+	err = d.createDiscoverDaemonSet(namespace, discoverImage)
+	if err != nil {
+		return fmt.Errorf("Error starting discover daemonset: %v", err)
+	}
+	return nil
+}
+
+func (d *Discover) createDiscoverDaemonSet(namespace, discoverImage string) error {
+	privileged := false
+	ds := &extensions.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: discoverDaemonsetName,
+		},
+		Spec: extensions.DaemonSetSpec{
+			UpdateStrategy: extensions.DaemonSetUpdateStrategy{
+				Type: extensions.RollingUpdateDaemonSetStrategyType,
+			},
+			Template: v1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app": discoverDaemonsetName,
+					},
+				},
+				Spec: v1.PodSpec{
+					ServiceAccountName: discoverDaemonsetName,
+					Containers: []v1.Container{
+						{
+							Name:  discoverDaemonsetName,
+							Image: discoverImage,
+							Args:  []string{"discover"},
+							SecurityContext: &v1.SecurityContext{
+								Privileged: &privileged,
+							},
+							VolumeMounts: []v1.VolumeMount{
+								{
+									Name:      "dev",
+									MountPath: "/dev",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "sys",
+									MountPath: "/sys",
+									ReadOnly:  true,
+								},
+								{
+									Name:      "udev",
+									MountPath: "/run/udev",
+									ReadOnly:  true,
+								},
+							},
+							Env: []v1.EnvVar{
+								k8sutil.NamespaceEnvVar(),
+								k8sutil.NodeEnvVar(),
+							},
+						},
+					},
+					Volumes: []v1.Volume{
+						{
+							Name: "dev",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/dev",
+								},
+							},
+						},
+						{
+							Name: "sys",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/sys",
+								},
+							},
+						},
+						{
+							Name: "udev",
+							VolumeSource: v1.VolumeSource{
+								HostPath: &v1.HostPathVolumeSource{
+									Path: "/run/udev",
+								},
+							},
+						},
+					},
+					HostNetwork: false,
+				},
+			},
+		},
+	}
+
+	// Add toleration if any
+	tolerationValue := os.Getenv(discoverDaemonsetTolerationEnv)
+	if tolerationValue != "" {
+		ds.Spec.Template.Spec.Tolerations = []v1.Toleration{
+			{
+				Effect:   v1.TaintEffect(tolerationValue),
+				Operator: v1.TolerationOpExists,
+				Key:      os.Getenv(discoverDaemonsetTolerationKeyEnv),
+			},
+		}
+	}
+
+	_, err := d.clientset.Extensions().DaemonSets(namespace).Create(ds)
+	if err != nil {
+		if !kserrors.IsAlreadyExists(err) {
+			return fmt.Errorf("failed to create rook-discover daemon set. %+v", err)
+		}
+		logger.Infof("rook-discover daemonset already exists")
+	} else {
+		logger.Infof("rook-discover daemonset started")
+	}
+	return nil
+
+}
+
+func ListDevices(context *clusterd.Context, namespace, nodeName string) (map[string][]sys.RawDevice, error) {
```

Comment:
Pardon the dumbness of the question, but i don't see this being used anywhere in the code right now. is this for the provisioner to gather devices on a host and prepare them for osds?

Reply:
yes, that's the 2nd step: rook operator will list devices using this function on each node and filter the devices. The filtered devices are then provisioned to become osds